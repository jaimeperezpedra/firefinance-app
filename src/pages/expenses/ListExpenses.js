import React from 'react';
import { Container, IconButton, Modal, Button, Box, TextField } from '@mui/material';
import DatePicker from '@mui/lab/DatePicker';
import AdapterDateFns from '@mui/lab/AdapterDateFns';
import LocalizationProvider from '@mui/lab/LocalizationProvider';
import { gql, useQuery, useMutation } from '@apollo/client';
import DeleteForeverIcon from '@mui/icons-material/DeleteForever';

import { SimpleTable, ModalBox } from '../../components'
import { CreateExpense } from './CreateExpense';


export const GET_ALL_EXPENSES = gql`
  query getExpenses($input: ExpenseFilter) {
    expenses(input: $input) {
      category
      price
      description
      currency
      id
    }
  }
`;

export const DELETE_EXPENSE_BY_ID = gql`
  mutation deleteExpense($input: ID!) {
    deleteExpense(input: $input)
  }
`;


export const ListExpenses = () => {
  const [date, setDate] = React.useState(new Date());
  const input = {
    month: date.getMonth() + 1,
    year: date.getFullYear(),
  }
  const { loading, data, error, refetch } = useQuery(GET_ALL_EXPENSES, 
    {
      variables: {
        input
      }
    });
  const [deleteExpenseByID] = useMutation(DELETE_EXPENSE_BY_ID);
  const [open, handleClose] = React.useState(false);
  const [idToDelete, setIdToDelete] = React.useState('');

  const columns = [
    {
      id: 'price',
      label: 'Price',
      render: ({value}) => {
        return `${value} â‚¬`
      },
    },
    {
      id: 'category',
      label: 'Category',
    },
    {
      id: 'description',
      label: 'Description',
    },
    {
      id: 'currency',
      label: 'Currency',
    },
    {
      id: 'custom',
      label: '',
      render: ({row}) => {
        return <IconButton onClick={() => openDeleteModal(row)}><DeleteForeverIcon color="action" /></IconButton>
      },
    }
  ];

  const deleteExpense = async () => {
    await deleteExpenseByID({
      variables: { input: idToDelete },
    });
    refetch();
    handleClose(false);
  }
  const openDeleteModal = ({id}) => {
    setIdToDelete(id);
    handleClose(true);
  }

  return (
    <Container sx={{ mt: 4, mb: 4 }}>
      <LocalizationProvider dateAdapter={AdapterDateFns}>
        <DatePicker
          views={['year', 'month']}
          label="Year and Month"
          value={date}
          onChange={(newDate) => {
            setDate(newDate);
          }}
          renderInput={(params) => <TextField {...params} helperText={null} />}
        />
      </LocalizationProvider>
      {loading && 'Loading...'}
      {error && 'Error...'}
      {data && <SimpleTable data={data.expenses} columns={columns} config={{dense: true}} />}

      <CreateExpense onCreated={refetch} />
      <Modal
        open={open}
        onClose={() => handleClose(false)}
      >
        {/* <div>hollo</div> */}
        <ModalBox>
          <h3>
            Are you sure you want to the delete the expense?
          </h3>
          <Box sx={ButtonsContainer}>
            <Button  variant="contained" onClick={() => handleClose(false)} color="primary">
              Close
            </Button>
            <Button  variant="contained" onClick={deleteExpense} color="error">
              Delete
            </Button>
          </Box>
        </ModalBox>
      </Modal>
    </Container>
  )
}

const ButtonsContainer = {
  display: 'flex',
  gap: 2,
  justifyContent: 'flex-end',
};
