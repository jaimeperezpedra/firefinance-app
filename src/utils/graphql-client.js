import { ApolloClient, from, HttpLink, InMemoryCache } from '@apollo/client';
import { onError } from '@apollo/client/link/error';

export const GRAPHQL_ENDPOINT_URI = '/graphql/';

export const errorLink = onError(({ graphQLErrors, networkError }) => {
  if (process.env.NODE_ENV === 'development') {
    if (graphQLErrors) {
      graphQLErrors.map(({ message }) =>
        console.error(`[GraphQL error]: Message: ${message}`)
      );
    }
    if (networkError) {
      console.error(`[Network error]: ${networkError}`);
    }
  }
});

export const defaultConfig = () => ({
  link: from([
    errorLink,
    new HttpLink({
      credentials: 'include',
      uri: new URL(GRAPHQL_ENDPOINT_URI, window.location),
      fetch: (...args) => fetch(...args),
    }),
  ]),
  cache: new InMemoryCache(),
  // Fetch policies say whether the client uses its cache, the network, or a
  // combination of the two. ApolloClient uses the cache too much by default.
  //
  // Generally speaking, the UI should display the latest data. However a
  // reasonable compromise between performance and freshness is to initially
  // display a cached value and immediately refresh from the network.
  //
  // Any query can override the defaults below.
  defaultOptions: {
    // `query` is the low-level API. The cache is ignored (but still updated) by default.
    query: {
      fetchPolicy: 'network-only',
    },
    // `watchQuery` is "observable" and used by query hooks (useQuery,
    // useLazyQuery, etc). The cache is used for the initial value but the value
    // may be updated by a background network call.
    watchQuery: {
      fetchPolicy: 'cache-and-network',
      nextFetchPolicy: 'cache-first',
    },
  },
});

export function createClient() {
  return new ApolloClient(defaultConfig());
}
