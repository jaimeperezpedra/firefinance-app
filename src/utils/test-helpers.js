import { ApolloProvider } from '@apollo/client';
import PropTypes from 'prop-types';
import React from 'react';

import { createClient } from './graphql-client';

const baseURL = 'http://ops.localhost';

export function definePathName(pathname) {
  global.jsdom?.reconfigure({
    url: `${baseURL}${pathname}`,
  });
}

// Wrap the child component tree with an ApolloProvider configured to use a
// default client.
export const TestingApolloProvider = ({ children, client }) => {
  if (!client) {
    client = createClient();
  }

  return <ApolloProvider client={client}>{children}</ApolloProvider>;
};

TestingApolloProvider.propTypes = {
  children: PropTypes.element.isRequired,
  client: PropTypes.object,
};
