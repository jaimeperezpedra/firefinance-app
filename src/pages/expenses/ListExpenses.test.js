import React from 'react';
import { render } from '@testing-library/react';
import { ListExpenses } from './ListExpenses.js';


import { TestingApolloProvider } from '../../utils/test-helpers'

describe('ListExpenses',() => {
  it('renders loading', () => {
    const { container } = render(<TestingApolloProvider><ListExpenses /></TestingApolloProvider>);
    expect(container).toMatchSnapshot();
  });
})