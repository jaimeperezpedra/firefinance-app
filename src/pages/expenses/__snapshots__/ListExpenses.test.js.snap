// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListExpenses renders loading 1`] = `
<div>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg css-1f2kw6w-MuiContainer-root"
  >
    <div
      class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for="mui-1"
        id="mui-1-label"
      >
        Year and Month
      </label>
      <div
        class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          aria-label="Choose date, selected date is Feb 26, 2022"
          aria-readonly="true"
          class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id="mui-1"
          readonly=""
          type="text"
          value="February 2022"
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-1z7n62"
          >
            <span>
              Year and Month
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    Loading...
    <button
      aria-label="add"
      class="MuiButtonBase-root MuiFab-root MuiFab-circular MuiFab-sizeLarge MuiFab-primary css-1euqxi8-MuiButtonBase-root-MuiFab-root"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="AddIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
